2018-11-05-language-server-protocol.md

```diff
@@ -0,0 +1,193 @@
+
+当用户执行 "跳转到定义（jump to definition）" 指令时，编辑器向 Swift language server 发送以下请求：
+
+```json
+{
+  "jsonrpc": "2.0",
+  "id": 1,
+  "method": "textDocument/definition",
+  "params": {
+    "textDocument": {
+      "uri": "file:///Users/NSHipster/Example.swift"
+    },
+    "position": {
+      "line": 1,
+      "character": 13
+    }
+  }
+}
+```
+
+收到这个请求后，Swift language server 使用 [SourceKit](https://github.com/apple/swift/tree/master/tools/SourceKit) 等编译器工具来标识相应的代码实体，并在代码的上一行找到其声明的位置。然后 language server 用以下消息进行响应:
+
+```json
+{
+  "jsonrpc": "2.0",
+  "id": 1,
+  "result": {
+    "uri": "file:///Users/NSHipster/Example.swift",
+    "range": {
+      "start": {
+        "line": 0,
+        "character": 6
+      },
+      "end": {
+        "line": 0,
+        "character": 12
+      }
+    }
+  }
+}
+```
+
+最后，编辑器导航到文件(在本例中，该文件已经打开)，将光标移动到该范围，并高亮显示出来。
+
+这种方法的美妙之处在于，编辑器完成所有这些操作时，除了 .swift 文件与 Swift 代码相关以外，对 Swift 编程语言一无所知。编辑器需要做的就是与 language server 对话并更新 UI。而且编辑器知道如何做到这一点后，就可以遵循相同的过程，与任何带有 language server 的语言所编写的代码进行交互。
+
+## Clang / LLVM 里的 Language Server Protocol
+
+如果你觉得之前的 _M + N_ 图有点眼熟，那可能是因为 LLVM 也采用了同样的方法。
+
+LLVM 的核心是中间表示（intermediate representation，IR）。LLVM 所支持的语言使用 _编译器前端（compiler frontend）_ 生成 IR，再使用 _编译器后端（compiler backend）_ 将 IR 生成所支持平台的机器码。
+
+{% asset lsp-llvm-ir.svg %}
+
+{% info %}
+如果你想了解 Swift 代码编译的更多细节，请查看 [我们关于 SwiftSyntax 的文章](https://nshipster.com/swiftsyntax/)。
+{% endinfo %}
+
+[Clang](https://clang.llvm.org) 是 C 语言的 LLVM 编译器前端。Swift 与 Objective-C 的互操作性（inter-operability）就是靠它实现的。在最近的 5.0.0 版本中，Clang 添加了一个名为 [Clangd](https://clang.llvm.org/extra/clangd.html) 的新工具，它是 LLVM 对 Language Server Protocol 的实现。
+
+2018 年 4 月，[苹果公司向 LLVM 邮件组宣布](http://lists.llvm.org/pipermail/cfe-dev/2018-April/057668.html)，将把开发的重心从 [libclang](https://clang.llvm.org/doxygen/group__CINDEX.html) 转向 Clangd，以其作为创建交互工具的主要方式。
+
+现在你可能会想，_“那又怎样？”_ 苹果公司是 LLVM 项目最重要的支持者之一，该项目创始人 Chris Lattner 已经在苹果公司工作了十多年。苹果公司决定从不透明的 Clang 工具切换到另一个，似乎是一个实现细节了(可以这么说)。
+
+这个官宣很有趣的一点是，Clangd 似乎完全是在苹果以外开发的，谷歌和其他公司也做出了重大贡献。这个官宣标志着未来工具开发方向的重大转变 —— 六个月后 Swift.org 论坛将证实这一点。
+
+## 苹果支持 Language Server Protocol 的潜在影响
+
+根据苹果公司 10 月份发布的 LSP 公告，我们预计在未来几周内（撰写本文时，最早 11 月中旬）将看到该项目的首批代码。
+
+要感受这些发展的全部影响还需要一些时间，但请相信我：你的耐心是值得的。我相信以下是 LSP 在未来几个月和几年将会发生的一些事情。
+
+### Swift 变成一种更加通用的编程语言
+
+虽然 Swift 主要用于 APP 开发，但它从一开始就被设计成一种功能强大的通用编程语言。在 [Swift for TensorFlow](https://www.tensorflow.org/swift/)、
+[SwiftNIO](https://github.com/apple/swift-nio) 和其他项目中，我们正开始看到 Swift 承诺的在 App Store 之外的使用。
+
+到目前为止，阻碍 Swift 被主流采用的最大因素之一是它对 Xcode 的依赖。
+
+人们会有很多质疑，当有那么多优秀的、门槛低很多的替代方案可选的情况下，为什么还要让 Web 开发者或机器学习工程师仅仅为了尝试 Swift 而去下载 Xcode？支持 Language Server Protocol 可以让苹果生态圈以外的人更容易地使用他们熟悉工具去感受 Swift。
+
+### Xcode 变得更好
+
+支持 LSP 不仅仅是让 Swift 在其他编辑器中运行地更好，Xcode 也将受益匪浅。
+
+看看苹果 Swift 项目负责人 Ted Kremenek 的 [这篇论坛帖子](https://forums.swift.org/t/new-lsp-language-service-supporting-swift-and-c-family-languages-for-any-editor-and-platform/17024/29):
+
+> @akyrtzi 所描述的 LSP 服务将比今天的 SourceKit 更强大。
```

Comment:
名字和原文不一样呀。

Reply:
因为我点进那个帖子后没有找到原文使用的 Argyrios，找到的是“the LSP service @akyrtzi is describing will be functionally more powerful than SourceKit is today.”。

Reply:
貌似又改动了？ https://nshipster.com/language-server-protocol/

Hey @mattt , is it supposed to be a @ link to GitHub, or just leave it in brackets like `[Argyrios]` ?

Reply:
`@akyrtzi` 是用户名，`Argyrios` 是人名，现在看起来译文中使用 `[Argyrios]` 更合适些。看 Mattt 怎么说。

Reply:
@croath .@mobilefellow is correct. The use of square brackets here ([ ]) is a journalistic convention that you use when you want to quote someone, but change content to provide missing context. In this case, I wanted to reference Argyrios by his name rather than username.

Reply:
👏